packages/react-dom-bindings/src/client/ReactDOMHostConfig.js

```diff
@@ -845,42 +859,210 @@ export function isHydratable(type: string, props: Props): boolean {
   }
 }
 
-export function canHydrateInstance(
+// In certain contexts, namely <body> and <head>, we want to skip past Nodes that are in theory
+// hydratable but do not match the current Fiber being hydrated. We track the hydratable node we
+// are currently attempting using this module global. If the hydration is unsuccessful Fiber will
+// call getLastAttemptedHydratable which uses this cursor to return the expected next
+// hydratable.
```

Comment:
This is quite a bug waiting to happen because it's very subtle when this needs to be reset and it'll likely not work the same when we switch to hydrating the commit phase. We need to have all state in FiberHydrationContext.

I get that you're trying to avoid returning multiple values but it might be better to just change the implementation to be closer to what the DOM actually needs.

Instead of getNextMatchingHydratableInstance, what about keeping getNextHydratableSibling for the iteration and just add something like shouldSkipHydratableInstance that's called for each one from the HydrationContext? Like that's probably more like what you would've written if you just wrote this inline without a consideration for HostConfigs.

Reply:
Did a refactor of the hydration context logic. Removes a bunch of branches but I'm not sure if it inlines better and has all the hydration state in FIber